### PR TITLE
Feat(Pipeline-run-logs): Add Pipeline Run Logs

### PR DESCRIPTION
**Pull request by abhinandan13jan**
Pipeline Run Logs
----------------------------------------
Screenshot from 2019-08-07 00-05-42
Screenshot from 2019-08-07 00-05-48
Screenshot from 2019-08-07 00-05-53
Screenshot from 2019-08-07 00-27-54

### CODE REVIEW

**Comment by abhinandan13jan** [comment]
> This looks awesome @abhinandan13jan !
> 
> Since you are already concatenating the task name, could we make the width of the log component wider?

updated

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+class PipelineRunLogs extends React.Component<PipelineRunLogsProps, PipelineRunLogsStates> {
+  constructor(props) {
+    super(props);
+    this.state = { activeItem: null };
+  }
+
+  componentDidMount() {
+    const { obj } = this.props;
+    const taskRunFromYaml =
+      obj.status && obj.status && obj.status.taskRuns ? obj.status.taskRuns : {};
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: `obj.status` is written twice.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    const taskRuns = this.getSortedTaskRun(taskRunFromYaml);
+    this.setState({ activeItem: taskRuns[taskRuns.length - 1] });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props !== nextProps) {
+      const { obj } = this.props;
+      const taskRunFromYaml =
+        obj.status && obj.status && obj.status.taskRuns ? obj.status.taskRuns : {};
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: same as above

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    this.setState({
+      activeItem: item.itemId,
+    });
+  };
+
+  render() {
+    const { obj } = this.props;
+    const { activeItem } = this.state;
+    const taskRunFromYaml =
+      obj.status && obj.status && obj.status.taskRuns ? obj.status.taskRuns : {};
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: same as above

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        isList: false,
+      },
+    ];
+    return (
+      <div className="odc-pipeline-run-logs">
+        <div className="col-xs-3 odc-pipeline-run-logs__task__list">
+          {taskCount > 0 ? (
+            <Nav onSelect={this.onNavSelect}>
+              <NavList className="odc-pipeline-run-logs__nav">
+                {taskRuns.map((a) => {
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: please change `a` to `task` or something meaningful

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import { StatusIcon } from '@console/shared';
+import { Firehose } from '@console/internal/components/utils';
+import { pipelineRunFilterReducer } from '../../utils/pipeline-filter-reducer';
+import { PipelineRun } from '../../utils/pipeline-augment';
+import './PipelineRunLogs.scss';
+
+interface PipelineRunLogsProps {
+  obj: PipelineRun;
+}
+interface PipelineRunLogsStates {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: nit:
```suggestion
interface PipelineRunLogsState {
```

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx (hunk 1 above)
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]:
```suggestion
      _.get(obj, ['status', 'taskRuns'], {})
```

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx (hunk 2 above)
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]:
```suggestion
      _.get(obj, ['status', 'taskRuns'], {})
```

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+  componentDidMount() {
+    const { obj } = this.props;
+    const taskRunFromYaml =
+      obj.status && obj.status && obj.status.taskRuns ? obj.status.taskRuns : {};
+    const taskRuns = this.getSortedTaskRun(taskRunFromYaml);
+    this.setState({ activeItem: taskRuns[taskRuns.length - 1] });
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props !== nextProps) {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: Better to compare `obj` here and make it more specific. Make sure that shallow compare is good enough as the condition in your case.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx (hunk 3 above)
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]:
```suggestion
      _.get(obj, ['status', 'taskRuns'], {})
```

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    const { obj } = this.props;
+    const { activeItem } = this.state;
+    const taskRunFromYaml =
+      obj.status && obj.status && obj.status.taskRuns ? obj.status.taskRuns : {};
+    const taskRuns = this.getSortedTaskRun(taskRunFromYaml);
+
+    const taskCount = taskRuns.length;
+    const resources = taskCount > 0 && [
+      {
+        name:
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]:
```suggestion
        name: _.get(taskRunFromYaml[activeItem],  ['status', 'podName'], ''),
```

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                {taskRuns.map((a) => {
+                  return (
+                    <NavItem
+                      key={a}
+                      itemId={a}
+                      isActive={activeItem === a}
+                      className="odc-pipeline-run-logs__nav__item"
+                    >
+                      <StatusIcon status={pipelineRunFilterReducer(obj.status.taskRuns[a])} />
+                      {
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: Remove extra braces.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                        <span className="odc-pipeline-run-logs__name__span">
+                          {taskRunFromYaml[a].pipelineTaskName.length < 20
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: Move this logic to a utility function that truncates the name.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+            <div className="odc-pipeline-run-logs__nav">No Task Runs Found</div>
+          )}
+        </div>
+        <div className="col-xs-9 odc-pipeline-run-logs__container">
+          {activeItem ? (
+            <Firehose resources={resources}>
+              <FetchTRPod />
+            </Firehose>
+          ) : obj.status &&
+            obj.status.conditions &&
```
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineRunLogs.tsx]: Use lodash `get` method to clean this up.

**Comment by rohitkrai03** [comment]
The last scenario which says `No task runs found` looks weird. Maybe just remove the terminal component that shows logs and just show that message? cc: @serenamarie125

**Comment by abhinandan13jan** [comment]
@rohitkrai03 @debsmita1 updated

**Pull request merged**